doc: add kb for manager crashing due to backing image eviction nil pointer (#1067)
longhorn/website: content/kb/troubleshooting-longhorn-manager-crashes-due-to-backing-image-eviction.md

---
title: "Troubleshooting: Longhorn Manager Crashes Due To Backing Image Eviction"
authors:
- "Jack Lin"
draft: false
date: 2025-03-07
versions:
- ">= v1.7.0"
categories:
- "longhorn manager"
---

## Applicable versions

Longhorn >= v1.7.0.

> 

Author: Yes you are right, we fixed it after.
I think v1.8.0 is also included?

2. The backing image status, however, is not immediately updated to reflect this deletion.
3. When the node or disk `EvictionRequested` flag is set to true, the function `syncBackingImageEvictionRequested()` is triggered to evict the backing image on that node or disk.
4. The function builds a lookup map of disks and their corresponding backing images based on the outdated backing image status.
5. When the function attempts to update `backingImage.spec.evictionRequested` for the evicted disk, it encounters a nil pointer, causing the longhorn-manager to crash.

## Workaround

To mitigate the issue and restore stability, follow these steps:

1. Set all EvictionRequested flags to false to prevent the eviction process from triggering the crash.
2. Wait for the disk cleanup process to complete and ensure that the backing image status is fully updated.
3. Once the cleanup is finished, resume the eviction process by setting EvictionRequested flags back to true as needed.

This workaround will allow the longhorn-manager to stabilize and prevent further crashes while ensuring that the eviction process can continue once the backing image status is consistent.

> 

Reviewer: Add 
```
## Fixed Versions

Longhorn v1.8.1+

## Reference

The related GitHub ticket
```

Author: added.

## Root Cause

The issue is caused by a race condition between the deletion of a disk in the backing image spec and the updating of its status, as explained in the following sequence:

1. Longhorn attempts to clean up unused backing images on a disk, which results in the disk being deleted from the backing image spec. The backing image status is not immediately updated to reflect this deletion.
2. When the node or disk `EvictionRequested` flag is set to true, the function `syncBackingImageEvictionRequested()` is triggered to evict the backing image from that node or disk.
3. The function builds a lookup map of disks and their corresponding backing images based on the outdated backing image status.
4. The function encounters a nil pointer when attempting to update `backingImage.spec.evictionRequested` for the evicted disk, causing Longhorn Manager to crash.

## Workaround

To mitigate the issue and restore stability, follow these steps:

1. Set all `EvictionRequested` flags to `false` to prevent the eviction process from triggering the crash.

> 2. Wait for the disk cleanup process to be completed, and ensure that the backing image status is fully updated.

Author: Fixed by adding
> You can verify this by using kubectl to retrieve all backing image CRs and checking if any disks appear in the status but are no longer present in the spec.

---
title: "Troubleshooting: Longhorn Manager Crashes Due To Backing Image Eviction"
authors:
- "Jack Lin"
draft: false
date: 2025-03-07
versions:

> - "v1.7.0-v1.7.3 and v1.8.0-v1.8.1"

Reviewer: Should be fixed in v1.8.1?
 
```suggestion
- "v1.7.0-v1.7.3 and v1.8.0"
```

Author: Oh yeah, I forgot we merged it before releasing 1.8.1...

Author: fixed.

categories:
- "longhorn manager"
---

## Applicable versions

v1.7.0-v1.7.3 and v1.8.0

## Symptoms

> Longhorn-Manager might repeatedly crash, causing disruptions in storage management operations. The logs indicate that the crash is due to a nil pointer dereference in the function `syncBackingImageEvictionRequested()`.

Reviewer 2: ```suggestion
Longhorn Manager repeatedly crashes, disrupting storage management operations. The logs indicate that the crashes are caused by a nil pointer dereference in the function `syncBackingImageEvictionRequested()`.
```

Author: fixed.

## Root Cause

> The issue occurs due to a race condition between the deletion of a disk in the backing image spec and the update of its corresponding status. The following sequence explains the root cause:
> 
> 1. Longhorn attempts to clean up unused backing images on a disk, which results in the disk being deleted from the backing image spec.
> 2. The backing image status, however, is not immediately updated to reflect this deletion.
> 3. When the node or disk `EvictionRequested` flag is set to true, the function `syncBackingImageEvictionRequested()` is triggered to evict the backing image on that node or disk.
> 4. The function builds a lookup map of disks and their corresponding backing images based on the outdated backing image status.
> 5. When the function attempts to update `backingImage.spec.evictionRequested` for the evicted disk, it encounters a nil pointer, causing the longhorn-manager to crash.

Reviewer 2: ```suggestion
The issue is caused by a race condition between the deletion of a disk in the backing image spec and the updating of its status, as explained in the following sequence:

1. Longhorn attempts to clean up unused backing images on a disk, which results in the disk being deleted from the backing image spec. The backing image status is not immediately updated to reflect this deletion.
1. When the node or disk `EvictionRequested` flag is set to true, the function `syncBackingImageEvictionRequested()` is triggered to evict the backing image from that node or disk.
1. The function builds a lookup map of disks and their corresponding backing images based on the outdated backing image status.
1. The function encounters a nil pointer when attempting to update `backingImage.spec.evictionRequested` for the evicted disk, causing Longhorn Manager to crash.
```

Author: fixed.

## Workaround

To mitigate the issue and restore stability, follow these steps:

> 1. Set all EvictionRequested flags to false to prevent the eviction process from triggering the crash.
> 2. Wait for the disk cleanup process to complete and ensure that the backing image status is fully updated.
> 3. Once the cleanup is finished, resume the eviction process by setting EvictionRequested flags back to true as needed.
> 
> This workaround will allow the longhorn-manager to stabilize and prevent further crashes while ensuring that the eviction process can continue once the backing image status is consistent.

Reviewer 2: ```suggestion
1. Set all `EvictionRequested` flags to `false` to prevent the eviction process from triggering the crash.
1. Wait for the disk cleanup process to be completed, and ensure that the backing image status is fully updated.
1. Resume the eviction process by setting the `EvictionRequested` flags back to `true` as necessary.

This workaround allows Longhorn Manager to stabilize itself, prevents further crashes, and ensures that the eviction process can continue once the backing image status is consistent.
```

Author: fixed.
